lib/utils/validation.js

```diff
@@ -113,3 +113,33 @@ export function validateValsAgainstVocab(delimited_string, field) {
   if (update_flag) return [true, value_array.join(';')];
   else return [true, false];
 }
+
+/**
+ *
+ * @param {Array<Array<any>>} columns to validate
+ * @return {Array<boolean>}
+ *         whether each row represents a unique combination of values
+ */
+export function validateUniqueValues(values) {
+  const results = [];
+  const longestColumn = Math.max.apply(
+    null,
+    values.map((col) => col.length)
+  );
+  for (let self = 0; self < longestColumn; self += 1) {
+    let valid = true;
+    const selfValue = values.map((col) => col[self]);
+    for (let other = 0; other < self; other += 1) {
```

Comment:
Is this going back over previous rows to spot a match to current one?  Or am I completely misreading this?

Reply:
The `validateUniqueValues` method takes an array of arrays. Each inner array represents a _column_ of data (if you look at the test it's a little confusing because it gets formatted in such a way that it looks like _rows_ of data). The algorithm is not terribly sophisticated. It first gets the length of each of the column arrays. The largest length represents the number of rows to check (just in case more rows have been entered for one column than the others). Then it proceeds as follows:
* For each row (this row number is referred to as `self`):
  * Make an array composed of each column's value in that row (`selfValue`)
  * For all previous rows checked (this row number is referred to as `other`)
    * Make an array composed of each column's value for that previous row (`otherValue`)
    * Check if all the values from the current row (`selfValue`) are the same as the previous row (`otherValue`) -- this is implemented as a call to `selfValue.every()` which might look a bit cryptic at first but it's necessary because `selfValue == otherValue` would do a _referential_ equality check on the two arrays and that will always be `false` in this case.
    * If `selfValue` and `otherValue` represent the same values, flag both the `self` row and `other` row as invalid.

I will add some additional comments to the code to explain what's going on. I can also see now that some of these variable names could also be changed to make it more clear.

Reply:
Is this doing a lot of looping as we get towards datasets with 1000+ rows? i.e. is comparison progression per row = 0,1,2,3,4,5,6,7,8,9... 1000+
and so sum of comparisons done up to row n =
0,1,3,6,10,15,21,28,36,45, ... 
(whatever that complexity is called?)

Reply:
Yeah it's O(n^2) complexity, but I'm not really sure that's a practical concern here. Nothing in the inner loop takes much time at all. On my machine validating 1,000 non-empty rows with 4 unique columns takes about 50ms. At 10,000 non-empty rows it takes about 2 seconds. I can take a look at what a hash-and-lookup-based implementation would look like and if it saves an appreciable amount of time.

Reply:
Yes, the hash approach crossed my mind: one loop through the rows. Assemble string values of each key into tab delimited string, lookup in dictionary assembled so far. If no value, its unique so far, enter into dictionary with value being row.  If collision, then resulting value is row to mark problem key fields in if they haven't already, as well as current row.

Reply:
The latest revision implements basically what you're saying. I used a Unicode null character (`\u0000`) instead of a tab to create a hash of the column values because I thought that would have an even lower chance of being encountered in user input. I tested up to 10,000 non-empty rows and found the validation to take less than 10ms.